### PR TITLE
[KERNEL32] Add NT6+ Thread_Local support

### PR DESCRIPTION
**Pull request by DibyaTheXPFan**
## Purpose

Windows XP/2K3 doesn't support thread_local storage , it will allow modern Browsers to run on ROS, if NT6+ API are provided by Compatibility layer.

## Proposed changes

Changes to TLS API 

## TODO
 NT6+ Compatibility layer for Firefox/Chrome

### CODE REVIEW

File: dll/win32/kernel32/client/thread.c [hunk 1]
```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
...
- * @implemented
- */
-DWORD
-WINAPI
-TlsAlloc(VOID)
-{
-    ULONG Index;
-    PTEB Teb;
-    PPEB Peb;
+// Modification by Dibyamartanda Samanta
```
Comment (SergeGautherie) [inline dll/win32/kernel32/client/thread.c:1095]: Unwanted comment.

File: dll/win32/kernel32/client/thread.c [hunk 2]
```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
...
-{
-    ULONG Index;
-    PTEB Teb;
-    PPEB Peb;
+// Modification by Dibyamartanda Samanta 
 
-    /* Get the PEB and TEB, lock the PEB */
-    Teb = NtCurrentTeb();
-    Peb = Teb->ProcessEnvironmentBlock;
+DWORD WINAPI TlsAlloc( void )
```
Comment (SergeGautherie) [inline dll/win32/kernel32/client/thread.c:1097]: Unwanted change.

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c [hunk 3]
```diff
@@ -3,68 +3,153 @@
...
+   PIO_STATUS_BLOCK       IoStatusBlock,
+   PVOID                  FileInformation,
+   ULONG                  Length,
+   FILE_INFORMATION_CLASS FileInformationClass);
+
+static pfnNtQueryInformationFile fnNtQueryInformationFile = nullptr;
+
+static BOOL bNtdllInited = false;
+
+static BOOL WINAPI GetFileInformationByHandleEx(
```
Comment (SergeGautherie) [inline dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:52]: `static`?

Comment (DibyaTheXPFan) [reply dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:52]: i wont compile withouth them for some reason on ROSBE

File: dll/win32/kernel32/client/thread.c (hunk 1 above)
Comment (DibyaTheXPFan) [reply dll/win32/kernel32/client/thread.c:1095]: Why? I modified it

Comment (whindsaks) [reply dll/win32/kernel32/client/thread.c:1095]: > I modified it

What would the code look like if everyone added their name to every little bit of the code they changed?

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c [hunk 4]
```diff
@@ -3,68 +3,153 @@
...
+   _In_  HANDLE hFile,
+   _In_  FILE_INFO_BY_HANDLE_CLASS FileInformationClass,
+   _Out_writes_bytes_(dwBufferSize) LPVOID lpFileInformation,
+   _In_  DWORD dwBufferSize)
+{
+   FILE_INFORMATION_CLASS NtFileInformationClass;
+   DWORD cbMinBufferSize;
+   BOOLEAN RestartScan = false;
+
+   switch (FileInformationClass) // Purosham Sahanu InformationClass
```
Comment (whindsaks) [inline dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:62]: Purosham Sahanu? This InformationClass is documented on MSDN...

Comment (DibyaTheXPFan) [reply dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:62]: My search engine Ecosia didnt link me to MSDN , no complains

**Review by binarymaster — CHANGES_REQUESTED** [review]
Hello,
Thanks for your interest in ReactOS!
We ask our contributors [to use their full name when committing](https://github.com/reactos/reactos/blob/master/CONTRIBUTING.md#rules-and-recommendations).
[Please amend your commit with your full name](https://stackoverflow.com/a/3042512), and update this PR.

To change this for future PR's, you can update it for globally (for the system):
```
git config --global user.name "Dibyamartanda Samanta"
git config --global user.email dibyasamanta90@hotmail.com
```

Or just for the current repository:
```
git config user.name "Dibyamartanda Samanta"
git config user.email dibyasamanta90@hotmail.com
```

Also please set up your GitHub account accordingly, so maintainers can use Squash & Merge button without getting the author field modified back to nickname by GitHub, it's done in two steps:

- Set your full name in [GitHub profile settings](https://github.com/settings/profile) (it should match your `user.name` setting)

Regards,

File: dll/win32/kernel32/client/thread.c (hunk 1 above)
Comment (binarymaster) [reply dll/win32/kernel32/client/thread.c:1095]: We use Git as version control system, your name will be stored in the commit author field anyway.

**Comment by HBelusca** [comment]
1. Please respect original code style;
2. Take the opportunity to use new SAL annotations;
3. New code for new exported functions require API tests.

**Comment by HBelusca** [comment]
> Windows XP/2K3 doesn't support thread_local storage

Then why are there functions called "TlsAlloc" etc. on WinXP or 2k3? What differ actually is how they function.
This blog suite: http://www.nynaeve.net/?p=189 may be a good reference (worth mentioning in the commit log).

File: dll/win32/kernel32/client/thread.c [hunk 5]
```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
     return TRUE;
 }
 
-/*
- * @implemented
- */
-DWORD
-WINAPI
-TlsAlloc(VOID)
```
Comment (HBelusca) [inline dll/win32/kernel32/client/thread.c:1100]: What about keeping the old implementation within some
`#if (_WIN32_WINNT < _WIN32_WINNT_VISTA) && (DLL_EXPORT_VERSION < _WIN32_WINNT_VISTA)`
block?

File: dll/win32/kernel32/client/thread.c [hunk 6]
```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
...
-            /* Clear the value. */
-            Teb->TlsExpansionSlots[Index] = 0;
-            Index += TLS_MINIMUM_AVAILABLE;
+            if (!NtCurrentTeb()->TlsExpansionSlots &&
+                !(NtCurrentTeb()->TlsExpansionSlots = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY,
+                                         8 * sizeof(peb->TlsExpansionBitmapBits) * sizeof(void*) )))
+            {
+                RtlClearBits( peb->TlsExpansionBitmap, index, 1 );
+                index = ~0U;
+                SetLastError( ERROR_NOT_ENOUGH_MEMORY );
```
Comment (HBelusca) [inline dll/win32/kernel32/client/thread.c:1119]: Could you compare this code of yours with the old one and tell us why those parts of the old code had to be changed?

File: dll/win32/kernel32/client/thread.c (hunk 5 above)
Comment (DibyaTheXPFan) [reply dll/win32/kernel32/client/thread.c:1100]: I will do it then , better if u can help me with that, excellent idea , both NT5 & NT6 version remains

Comment (HBelusca) [reply dll/win32/kernel32/client/thread.c:1100]: For that, look at the original code and yours, and see where there are actual functional differences.

File: dll/win32/kernel32/client/thread.c (hunk 6 above)
Comment (DibyaTheXPFan) [reply dll/win32/kernel32/client/thread.c:1119]: I re implemented it , i will compare it and come back

File: dll/win32/kernel32/client/thread.c [hunk 7]
```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
...
+        NtCurrentTeb()->TlsSlots[index] = value;
     }
-
-    /* Do we not have expansion slots? */
-    if (!Teb->TlsExpansionSlots)
+    if(index >= TLS_MINIMUM_AVAILABLE)
     {
-        /* Get the PEB lock to see if we still need them */
-        RtlAcquirePebLock();
-        if (!Teb->TlsExpansionSlots)
```
Comment (GeoB99) [inline dll/win32/kernel32/client/thread.c:1304]: Could you explain how does removing the PEB lock actually improve anything? If two or more concurrent threads are trying to allocate expansion slots from heap, would not that lead to erratic behaviors?

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c [hunk 8]
```diff
@@ -3,68 +3,153 @@
...
-        return FALSE;
-    }
-    return TRUE;
+   union {
+      NTSTATUS Status;
+      PVOID Pointer;
+   } DUMMYUNIONNAME;
+
+   ULONG_PTR Information;
+} IO_STATUS_BLOCK, *PIO_STATUS_BLOCK;
```
Comment (GeoB99) [inline dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:24]: Why are you duplicating the structure? `iotypes.h` from NDK already has it declared and the old code was even using it.

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c (hunk 3 above)
Comment (GeoB99) [reply dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:52]: What compilation error you're getting? Probably this function needs a prototype declaration in another place alongside PSDK. Declaring it static makes the function unusable to the outside callers....

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c [hunk 9]
```diff
@@ -3,68 +3,153 @@
...
+         nullptr,
+         nullptr,
+         &IoStatusBlock,
+         lpFileInformation,
+         dwBufferSize,
+         NtFileInformationClass,
+         false,
+         nullptr,
+         RestartScan
+      );
```
Comment (GeoB99) [inline dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:118]: Seriously, why? The previous code could invoke `NtQueryDirectoryFile` syscall just fine, why are you complicating things even further by loading the NTDLL library by yourself?

**Review by GeoB99 — CHANGES_REQUESTED** [review]

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c (hunk 9 above)
Comment (DibyaTheXPFan) [reply dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:118]: Well i forgot it's rosbe and got NDK, 
in Visual Studio , you need to call it yourself

**Comment by SergeGautherie** [comment]
This PR looks like a Draft to me...

File: dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c (hunk 9 above)
Comment (tkreuzer) [reply dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c:118]: No you don't, you just need to create an import library for ntdll.

File: dll/win32/kernel32/client/thread.c (hunk 6 above)
Comment (tkreuzer) [reply dll/win32/kernel32/client/thread.c:1119]: And why did you "reimplement" it? It's functionally almost identical and it doesn't fix anything, it rather introduces a bug.

**Review by tkreuzer — CHANGES_REQUESTED** [review]
This doesn't add or implement anything, it just uselessly rewrites exisiting code, reducing readability and introducing hacks and bugs. Whether this even fixes or improves anything is hard to tell, because 90% of it is useless.

**Review by DarkFire01 — CHANGES_REQUESTED** [review]
The NT6 changes to Thread Local Storage go way deeper then kernel32. It requires creating an expandable list within ntdll and changing the pointers for the active thread using a job in ntoskrnl when a dll is loaded outside of the initial import list.
 some more information can be gathered here:
[MSDN]( https://learn.microsoft.com/en-us/cpp/parallel/thread-local-storage-tls?view=msvc-170) 
[A VERY good outline](http://www.nynaeve.net/?p=189)
 (Thank you Hermes for the resources as always :))
 
 You were right about this being a key factor in browsers working. 
 Firefox 54 for instance introduced the requirement for this feature for the browser in order to not instantly crash on ReactOS (or of course Windows).
 You can also write a test case for this situation which is how I was investigating this subject originally.
 
As the other devs have said the changes here don't seem to make an observable positive change.
However, given our interactions up till now. I won't be the one to close this.

**Comment by SergeGautherie** [comment]
Rebase needed, after #5802.

**Pull request closed without merging**